hax/test/integration/test_motr.py

```diff
@@ -71,6 +73,109 @@ def fn(trace: Invocation) -> bool:
     return fn
 
 
+def node_failed() -> TraceMatcher:
+    '''
+    Returns true if the second argument is a array to pointer to HaNoteStruct
+    and it contains 5 elements and that structure brings M0_NC_FAILED state.
+    In case of node failure we should see 5 elements(1 IO service, 1 node fid,
+    1 encl fid, 2 ctrls) in array.
+
+    It makes sense to use this matcher together with
+    `tr_method('ha_broadcast')`
+    '''
+    def fn(trace: Invocation) -> bool:
+        if len(trace.args) < 2:
+            return False
+        notes = trace.args[1]
+        if (not isinstance(notes, ctypes.Array) or
+                len(notes) != 5):
+            return False
+
+        fids = [FidStruct(0x7200000000000001, 0x15),
+                FidStruct(0x6e00000000000001, 0x3),
+                FidStruct(0x6500000000000001, 0x4),
+                FidStruct(0x6300000000000001, 0x5),
+                FidStruct(0x6300000000000001, 0x6)]
+
+        for _, note in enumerate(notes):
+            if not isinstance(note, HaNoteStruct):
+                return False
+            state: int = note.no_state
+            if state != HaNoteStruct.M0_NC_FAILED:
+                return False
+            fid = note.no_id
+            bMatch = False
+            for _, out_fid in enumerate(fids):
+                if (fid.f_container == out_fid.f_container and
+                        fid.f_key == out_fid.f_key):
+                    bMatch = True
+                    break
+
+            if not bMatch:
+                return False
+
+        return True
+
+    return fn
+
+
+def node_not_failed() -> TraceMatcher:
```

Comment:
Not sure that it works.

You see, TraceMatcher is a function applied to one particular invocation trace. TraceMatcher "doesn't see" all the traces at the same time, it "sees" just one trace at a time. TraceMatcher can say whether this particular trace matches the criteria. But it can't actually say whether ALL traces match or don't match.

So I would go another way.
1. Implement a matcher that matches a node in the traces ("a trace contains FID with node FID")
2. And then implement `not exist` operation (presumably at the level of AssertionPlan class).

So effectively you could build a predicate "the exist no invocation trace that has a node fid"

Reply:
@knekrasov I think we are dealing with single trace only(if I am not wrong).
We are checking if Fid is present in that trace args or not.
Let me know your inputs on this

Reply:
There is no proof that there is only one trace :-)

In other words, if by any chance or because of a new piece of functionality there will be 2 traces, your test should still be valid. If the second or the third trace contains a node, then the test should fail